l10n_es_genci_purchase/models/purchase_order.py

```diff
@@ -0,0 +1,186 @@
+    def _compute_genci_company(self):
+        for order in self:
+            order.genci_company = bool(order.company_id.genci_enable)
+
+    @api.onchange("partner_id")
+    def _onchange_partner_genci(self):
+        for order in self:
+            if not order.partner_id:
+                order.is_genci = False
+                return
+
+            if order.partner_id.genci_subject:
+                order.is_genci = True
+            else:
+                order.is_genci = False
+                genci_product = self.env.ref(
+                    "l10n_es_genci_account.product_genci_service"
+                )
+                order.order_line = order.order_line.filtered(
+                    lambda l: l.product_id != genci_product
+                )
+
+    @api.constrains("order_line", "is_genci")
+    def _check_genci_rules_dates(self):
+        for order in self.filtered(lambda o: o.is_genci and o.partner_id.genci_subject):
+            date_to_check = fields.Date.to_date(
+                order.date_order or fields.Datetime.now()
+            )
+            genci_lines = order.order_line.filtered(
+                lambda l: l.product_id.genci_subject == "yes"
+                and (
+                    l.product_id.genci_rule_id
+                    or l.product_id.product_tmpl_id.genci_rule_id
+                )
+            )
+            for line in genci_lines:
+                rule = (
+                    line.product_id.genci_rule_id
+                    or line.product_id.product_tmpl_id.genci_rule_id
+                )
+                rule_from = (
+                    fields.Date.to_date(rule.date_from) if rule.date_from else None
+                )
+                rule_to = fields.Date.to_date(rule.date_to) if rule.date_to else None
+                if (rule_from and date_to_check < rule_from) or (
+                    rule_to and date_to_check > rule_to
+                ):
+                    raise UserError(
+                        _(
+                            "No GENCI rule is valid for the product '%(product)s' "
+                            "on the order date (%(date)s)."
+                        )
+                        % {
+                            "product": line.product_id.display_name,
+                            "date": date_to_check,
+                        }
+                    )
+
+    def manage_genci_order_lines(self):
+        genci_product = self.env.ref("l10n_es_genci_account.product_genci_service")
+
+        for order in self:
+            existing = order.order_line.filtered(
+                lambda l: l.product_id == genci_product
+            )
+            if existing:
+                existing.unlink()
+            if order.fiscal_position_id:
+                apply_genci = order.fiscal_position_id.genci_subject
+            else:
+                apply_genci = order.is_genci and order.partner_id.genci_subject
+
+            if not apply_genci:
```

Comment:
This shouldn't be possible, and if it were, it should result in a controlled error.